actors/builtin/market/market_test.go

```diff
@@ -1 +1,347 @@
+	"github.com/filecoin-project/specs-actors/actors/util/adt"
+	"github.com/filecoin-project/specs-actors/support/mock"
+	tutil "github.com/filecoin-project/specs-actors/support/testing"
+)
+
+func TestMarketActor(t *testing.T) {
+	marketActor := tutil.NewIDAddr(t, 100)
+	owner := tutil.NewIDAddr(t, 101)
+	provider := tutil.NewActorAddr(t, "provider")
+	providerIdAddr := tutil.NewIDAddr(t, 102)
+	worker := tutil.NewIDAddr(t, 103)
+	client := tutil.NewIDAddr(t, 104)
+	var st market.State
+
+	setup := func() (*mock.Runtime, *marketActorTestHarness) {
+		builder := mock.NewBuilder(context.Background(), marketActor).
+			WithCaller(builtin.SystemActorAddr, builtin.InitActorCodeID).
+			WithActorType(owner, builtin.AccountActorCodeID).
+			WithActorType(worker, builtin.AccountActorCodeID).
+			WithActorType(providerIdAddr, builtin.StorageMinerActorCodeID).
+			WithActorType(client, builtin.AccountActorCodeID)
+
+		rt := builder.Build(t)
+
+		rt.AddIDAddress(owner, owner)
+		rt.AddIDAddress(worker, worker)
+		rt.AddIDAddress(client, client)
+		rt.AddIDAddress(provider, providerIdAddr)
+
+		actor := marketActorTestHarness{t: t}
+		actor.constructAndVerify(rt)
+
+		return rt, &actor
+	}
+
+	t.Run("simple construction", func(t *testing.T) {
+		actor := market.Actor{}
+		receiver := tutil.NewIDAddr(t, 100)
+		builder := mock.NewBuilder(context.Background(), receiver).
+			WithCaller(builtin.SystemActorAddr, builtin.InitActorCodeID)
+
+		rt := builder.Build(t)
+
+		rt.ExpectValidateCallerAddr(builtin.SystemActorAddr)
+
+		params := adt.EmptyValue{}
+		ret := rt.Call(actor.Constructor, &params).(*adt.EmptyValue)
+
+		assert.Equal(t, adt.EmptyValue{}, *ret)
+		rt.Verify()
+
+		store := adt.AsStore(rt)
+
+		emptyMap, err := adt.MakeEmptyMap(store)
+		assert.NoError(t, err)
+
+		emptyArray, err := adt.MakeEmptyArray(store)
+		assert.NoError(t, err)
+
+		emptyMultiMap, err := market.MakeEmptySetMultimap(store)
+		assert.NoError(t, err)
+
+		var state market.State
+		rt.GetState(&state)
+
+		assert.Equal(t, emptyArray.Root(), state.Proposals)
+		assert.Equal(t, emptyArray.Root(), state.States)
+		assert.Equal(t, emptyMap.Root(), state.EscrowTable)
+		assert.Equal(t, emptyMap.Root(), state.LockedTable)
+		assert.Equal(t, abi.DealID(0), state.NextID)
+		assert.Equal(t, emptyMultiMap.Root(), state.DealIDsByParty)
+	})
+
+	t.Run("AddBalance", func(t *testing.T) {
+		t.Run("adds to provider escrow funds", func(t *testing.T) {
+			testCases := []struct {
+				delta int64
+				total int64
+			}{
+				{10, 10},
+				{20, 30},
+				{40, 70},
+			}
+
+			// Test adding provider funds from both worker and owner address
+			for _, callerAddr := range []address.Address{owner, worker} {
+				rt, actor := setup()
+
+				for _, tc := range testCases {
+					rt.SetCaller(callerAddr, builtin.AccountActorCodeID)
+					rt.SetReceived(abi.NewTokenAmount(tc.delta))
+					actor.expectProviderControlAddress(rt, providerIdAddr, owner, worker)
```

Comment:
~This should also expect a `ValidateImmediateCallerIs(owner, worker)`. I'm confused as to why this test didn't fail due to the actor making that call, but there being no expectation for it.~

Edit: I see it's inside this method call. Please rename something like `expectValidateCallerIsProviderControlAddress` or `expectProviderControlAddressAndValidateCaller` or pull that caller validation out and make it explicit here.